src/validators/enum_.rs

```diff
@@ -0,0 +1,212 @@
+
+use super::is_instance::class_repr;
+use super::literal::{expected_repr_name, LiteralLookup};
+use super::{BuildValidator, CombinedValidator, DefinitionsBuilder, Exactness, ValidationState, Validator};
+
+#[derive(Debug, Clone)]
+pub struct BuildEnumValidator;
+
+impl BuildValidator for BuildEnumValidator {
+    const EXPECTED_TYPE: &'static str = "enum";
+
+    fn build(
+        schema: &Bound<'_, PyDict>,
+        _config: Option<&Bound<'_, PyDict>>,
+        _definitions: &mut DefinitionsBuilder<CombinedValidator>,
+    ) -> PyResult<CombinedValidator> {
+        let members: Bound<PyList> = schema.get_as_req(intern!(schema.py(), "members"))?;
+        if members.is_empty() {
+            return py_schema_err!("`members` should have length > 0");
+        }
+
+        let py = schema.py();
+        let value_str = intern!(py, "value");
+        let expected: Vec<(Bound<'_, PyAny>, PyObject)> = members
+            .iter()
+            .map(|v| Ok((v.getattr(value_str)?, v.into())))
+            .collect::<PyResult<_>>()?;
+
+        let repr_args: Vec<String> = expected
+            .iter()
+            .map(|(k, _)| k.repr()?.extract())
+            .collect::<PyResult<_>>()?;
+
+        let class: Bound<PyType> = schema.get_as_req(intern!(py, "cls"))?;
+        let class_repr = class_repr(schema, &class)?;
+
+        let lookup = LiteralLookup::new(py, expected.into_iter())?;
+
+        macro_rules! build {
+            ($vv:ty, $name_prefix:literal) => {
+                EnumValidator {
+                    phantom: PhantomData::<$vv>,
+                    class: class.clone().into(),
+                    lookup,
+                    missing: schema.get_as(intern!(py, "missing"))?,
+                    expected_repr: expected_repr_name(repr_args, "").0,
+                    class_repr: class_repr.clone(),
+                    name: format!("{}[{class_repr}]", $name_prefix),
+                }
+            };
+        }
+
+        let sub_type: Option<String> = schema.get_as(intern!(py, "sub_type"))?;
+        match sub_type.as_deref() {
+            Some("int") => Ok(CombinedValidator::IntEnum(build!(IntEnumValidator, "int-enum"))),
+            Some("str") => Ok(CombinedValidator::StrEnum(build!(StrEnumValidator, "str-enum"))),
+            Some("float") => Ok(CombinedValidator::FloatEnum(build!(FloatEnumValidator, "float-enum"))),
+            Some(_) => py_schema_err!("`sub_type` must be one of: 'int', 'str', 'float' or None"),
+            None => Ok(CombinedValidator::PlainEnum(build!(PlainEnumValidator, "enum"))),
+        }
+    }
+}
+
+pub trait EnumValidateValue: std::fmt::Debug + Clone + Send + Sync {
+    fn validate_value<'py, I: Input<'py> + ?Sized>(
+        py: Python<'py>,
+        input: &I,
+        lookup: &LiteralLookup<PyObject>,
+        strict: bool,
+    ) -> ValResult<Option<PyObject>>;
+}
+
+#[derive(Debug, Clone)]
+pub struct EnumValidator<T: EnumValidateValue> {
+    phantom: PhantomData<T>,
```

Comment:
You could have also just gone for `T` rather than `PhantomData<T>` and made the `validate_value` trait method take `&self`; it'd probably optimize to the same machine code. But this is also totally fine 👍

Reply:
I had that, then changed it to this 🙃.